source/base/DateTime.ooc

```diff
@@ -14,22 +14,246 @@
+		result = result replaceAll("%ss", "%d" format(data second))
+		result = result replaceAll("%zzzz", "%d" format(data millisecond))
+		result
+	}
+
+	compareTo: func (other: This) -> Order {
+		if (this ticks > other ticks)
+			Order greater
+		else if (this ticks < other ticks)
+			Order less
+		else
+			Order equal
+	}
+
+	operator - (other: This) -> TimeSpan {
+		TimeSpan new(this ticks as Int64 - other ticks as Int64)
+	}
+	operator + (span: TimeSpan) -> This {
+		This new(this ticks as Int64 + span ticks)
+	}
+	operator - (span: TimeSpan) -> This {
+		This new(this ticks as Int64 - span ticks)
+	}
+	operator == (other: This) -> Bool {
+		this compareTo(other) == Order equal
+	}
+	operator != (other: This) -> Bool {
+		this compareTo(other) != Order equal
+	}
+	operator < (other: This) -> Bool {
+		this compareTo(other) == Order less
+	}
+	operator <= (other: This) -> Bool {
+		this compareTo(other) != Order greater
+	}
+	operator > (other: This) -> Bool {
+		this compareTo(other) == Order greater
+	}
+	operator >= (other: This) -> Bool {
+		this compareTo(other) != Order less
+	}
+
+	/* number of days in year ( non-leap ) */
+	DaysInYear: static const Int = 365
+	DaysInFourYears: static const Int = 3 * 365 + 366
+	TicksPerMillisecond: static const Int64 = 1000
+	TicksPerSecond: static const Int64 = TicksPerMillisecond * 1000
+	TicksPerMinute: static const Int64 = TicksPerSecond * 60
+	TicksPerHour: static const Int64 = TicksPerMinute * 60
+	TicksPerDay: static const Int64 = TicksPerHour * 24
+	TicksPerWeek: static const UInt64 = TicksPerDay * 7
+	TicksPerFourYears: static const Int64 = DaysInFourYears * TicksPerDay
+	/* default date/time printing format */
+	DefaultFormat: static const String = "%yyyy-%MM-%dd %hh:%mm:%ss::%zzzz"
+
+	_ticksToDateTimeHelper: static func (totalTicks: Int64) -> DateTimeData {
+		result := DateTimeData new()
+		fourYearBlocks := totalTicks / This TicksPerFourYears
+		year := 4 * fourYearBlocks
+		ticksLeft := totalTicks - fourYearBlocks * This TicksPerFourYears
+		for (y in year + 1 .. year + 5) {
+			t := This ticksInYear(y)
+			if (ticksLeft < t) {
+				year = y
+				break
+			} else {
+				ticksLeft -= t
+			}
+		}
+		month := 0
+		for (m in 1 .. 13) {
+			t := This ticksInMonth(year, m)
+			if (ticksLeft < t) {
+				month = m
+				break
+			} else {
+				ticksLeft -= t
+			}
+		}
+		days := ticksLeft / This TicksPerDay
+		ticksLeft -= days * This TicksPerDay
+		hour := ticksLeft / This TicksPerHour
+		ticksLeft -= hour * This TicksPerHour
+		minute := ticksLeft / This TicksPerMinute
+		ticksLeft -= minute * This TicksPerMinute
+		second := ticksLeft / This TicksPerSecond
+		ticksLeft -= second * This TicksPerSecond
+		millisecond := ticksLeft / This TicksPerMillisecond
+		result year = year
+		result month = month
+		result day = days + 1
+		result hour = hour
+		result minute = minute
+		result second = second
+		result millisecond = millisecond
+		result
```

Comment:
If you give `DateTimeData` a constructor, you won't need to write all this on separate lines.